### PR TITLE
Added ability to enable/disable scheduler and webserver 

### PR DESCRIPTION
**Pull request by kajanajustin**
Adding the ability to enable and disable the scheduler and webserver if you just want to run just workers in kubernetes. This change adds the enabled boolean in both the scheduler and webserver sections of the values.yaml file

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by cnkumar20 — APPROVED** [review]

**Comment by potiuk** [comment]
I am not sure we want it. If you reallly want to do worker-only deployment, you should create your own helm/resources -that will be way simpler and our chart is not meant to be as versatile to support such cases. cc: @jedcunningham @ephraimbuddy - WDYT?

**Comment by jedcunningham** [comment]
@potiuk I've actually come around to this idea over time. You can sorta do it today with scaling those components to 0, but you end up with all that stuff in k8s anyways. Plus it also can help with the worker queues problem (#34219) - not super gracefully, mind you, but better than roll your own certainly, as you can just use multiple values files to keep it sane.

I don't see this as being materially different than some of the other advanced stuff we expose to our users already.

**Comment by potiuk** [comment]
> @potiuk I've actually come around to this idea over time. You can sorta do it today with scaling those components to 0, but you end up with all that stuff in k8s anyways. Plus it also can help with the worker queues problem (#34219) - not super gracefully, mind you, but better than roll your own certainly, as you can just use multiple values files to keep it sane.
> 
> I don't see this as being materially different than some of the other advanced stuff we expose to our users already.

Fine for me then - if we are ready to support those cases :). I actually like it if we do it because this might help in some of the future descriptions of multi-tenant deployment. I initially thought that we do not want to make anything about Hel Chart with multitenancy, but actually now that you wrote it - it seems pretty plausible for anyone doing future multi-tenancy setup to do it by composing multiple charts. So yeah. I see how it can be useful if we want to support it.

It would be good however maybe in this case to add a bit of documentation about that case - for example describe a few deployments (like the multiple queue case) to explain users what our intention is with those parameters. @kajanajustin - would you be open to add a short description of potential use cases for those parameters ? Maybe in "Production Guide" part of the docs: https://airflow.apache.org/docs/helm-chart/stable/production-guide.html ?

**Comment by jedcunningham** [comment]
We also need test coverage for this 👍.

**Comment by kajanajustin** [comment]
@potiuk @jedcunningham Let me know what you guys think about the doc edits I made. Let me know if I need to add any additional descriptions

**Comment by potiuk** [comment]
Minimal, but good enough for me :) . Still a few tests would be needed to approve that one.

**Comment by kajanajustin** [comment]
@potiuk anything else needed from me at this time? I assume you guys have the testing handled from here

**Comment by potiuk** [comment]
> @potiuk anything else needed from me at this time? I assume you guys have the testing handled from here

No - unit tests are always part of the PRs. We have separately run tests for the helm chart in `helm_tests` folder which are basically testing that when you set the right parameter, the resources are produced properly after passing throuhg helm.

Our contributig guide is currently not properly rendered by github, but I have a PR in progress where all the contributing docs ( are nicely split to smaller docs, so you can see the docs  nicely rendering here https://github.com/apache/airflow/blob/refactor-contributing-docs/contributing-docs/testing/helm_unit_tests.rst and you can look at other tests in `helm_tests` and add your tests in similar way.

**Comment by potiuk** [comment]
also you have static checks failing and you should fix them. For that you can take a look at this split doc in my PR https://github.com/apache/airflow/blob/refactor-contributing-docs/contributing-docs/08_static_code_checks.rst

**Review by hussein-awala — COMMENTED** [review]
Looks good, just need a test per resource to ensure that it is not created when the enabled value is false.

**Comment by kajanajustin** [comment]
@hussein-awala how do we go about testing that?

**Comment by potiuk** [comment]
> No - unit tests are always part of the PRs. We have separately run tests for the helm chart in `helm_tests` folder which are basically testing that when you set the right parameter, the resources are produced properly after passing throuhg helm.
> 
> Our contributig guide is currently not properly rendered by github, but I have a PR in progress where all the contributing docs ( are nicely split to smaller docs, so you can see the docs nicely rendering here https://github.com/apache/airflow/blob/refactor-contributing-docs/contributing-docs/testing/helm_unit_tests.rst and you can look at other tests in `helm_tests` and add your tests in similar way.

That's what I described in https://github.com/apache/airflow/pull/36991#issuecomment-1908828287

**Comment by kajanajustin** [comment]
@hussein-awala @potiuk added additional unit tests, let me know if there's anything else you guys need to see

**Review by potiuk — APPROVED** [review]

**Review by dstandish — APPROVED** [review]

**Comment by kajanajustin** [comment]
@potiuk static checks and the test_can_be_disabled for the scheduler were failing so I added a new class for the function. Let me know if you think everything looks good so we can merge this in

**Review by jedcunningham — APPROVED** [review]

File: helm_tests/airflow_core/test_scheduler.py [hunk 1]
```diff
@@ -901,3 +902,19 @@ def test_overridden_automount_service_account_token(self):
             show_only=["templates/scheduler/scheduler-serviceaccount.yaml"],
         )
         assert jmespath.search("automountServiceAccountToken", docs[0]) is False
+
+class TestSchedulerCreation:
+    """Tests scheduler deployment creation."""
+
```
Comment (eladkal) [inline helm_tests/airflow_core/test_scheduler.py]:
```suggestion

```

to fix static checks

File: helm_tests/airflow_core/test_scheduler.py [hunk 2]
```diff
@@ -901,3 +902,18 @@ def test_overridden_automount_service_account_token(self):
             show_only=["templates/scheduler/scheduler-serviceaccount.yaml"],
         )
         assert jmespath.search("automountServiceAccountToken", docs[0]) is False
+
```
Comment (eladkal) [inline helm_tests/airflow_core/test_scheduler.py:904]:
```suggestion


```

File: helm_tests/airflow_core/test_scheduler.py [hunk 3]
```diff
@@ -901,3 +902,19 @@ def test_overridden_automount_service_account_token(self):
...
+
+class TestSchedulerCreation:
+    """Tests scheduler deployment creation."""
+    def test_can_be_disabled(self):
```
Comment (jedcunningham) [inline helm_tests/airflow_core/test_scheduler.py:909]:
```suggestion

    def test_can_be_disabled(self):
```

File: helm_tests/airflow_core/test_scheduler.py [hunk 4]
```diff
@@ -39,6 +39,7 @@ class TestScheduler:
             ("LocalExecutor", False, "Deployment"),
         ],
     )
+
```
Comment (jedcunningham) [inline helm_tests/airflow_core/test_scheduler.py]:
```suggestion
```

File: chart/templates/scheduler/scheduler-deployment.yaml [hunk 5]
```diff
@@ -20,7 +20,8 @@
 ################################
 ## Airflow Scheduler Deployment/StatefulSet
 #################################
-
+# Are we using a scheduler in k8s?
```
Comment (amoghrajesh) [inline chart/templates/scheduler/scheduler-deployment.yaml]: This comment raises a question, can we instead have a comment saying: `Condition for cases where we run only workers in K8s`

File: chart/templates/webserver/webserver-deployment.yaml [hunk 6]
```diff
@@ -20,6 +20,7 @@
 ################################
 ## Airflow Webserver Deployment
 #################################
+{{- if .Values.webserver.enabled }}
```
Comment (amoghrajesh) [inline chart/templates/webserver/webserver-deployment.yaml:23]: Same comment as above

**Review by amoghrajesh — COMMENTED** [review]
Generally looks good, wondering if we should add tests to the other k8s objects that are deployed if shceduler / webserver flags are turned on.

File: chart/templates/scheduler/scheduler-deployment.yaml (hunk 5 above)
Comment (jedcunningham) [reply chart/templates/scheduler/scheduler-deployment.yaml]: It honestly may be best to just remove the comment. I think its pretty obvious what the condition checks for, regardless of why the user chooses it.

**Comment by kajanajustin** [comment]
@jedcunningham what else needs to be updated here? I see some checks are still failing anything needed from me to merge this?

**Comment by jedcunningham** [comment]
I'm pretty sure those failures aren't related to your change, but I need to look into a little closer before I can ignore them. Also, we need to handle the comment discussion.

**Comment by potiuk** [comment]
Rebvased it. Those were issues connected with yesterday's "google-core-api" relase brekaing compatibility (we workarounded it but also they fixed it this morning with 2.16.1 release after we raised the bug.

File: chart/templates/scheduler/scheduler-deployment.yaml (hunk 5 above)
Comment (jedcunningham) [reply chart/templates/scheduler/scheduler-deployment.yaml]: @amoghrajesh, what do you think about just getting rid of this comment instead?

Comment (amoghrajesh) [reply chart/templates/scheduler/scheduler-deployment.yaml]: Go for it @kajanajustin , it makes sense to me not to have it at all.

Comment (kajanajustin) [reply chart/templates/scheduler/scheduler-deployment.yaml]: @amoghrajesh kk getting rid of it

**Comment by kajanajustin** [comment]
@jedcunningham comment has been removed can we merge this in

**Review by enchant3dmango — APPROVED** [review]

**Review by amoghrajesh — APPROVED** [review]

**Comment by kajanajustin** [comment]
@potiuk @jedcunningham dont want this to get stale and want to use this on the next helm chart release. 

Can we merge this in? Looks like everyone has approved who had comments before

**Review by potiuk — APPROVED** [review]

**Comment by jedcunningham** [comment]
Thanks @kajanajustin! Congrats on your first commit 🎉

**Comment by kajanajustin** [comment]
@jedcunningham thank you for all your help!

File: chart/templates/webserver/webserver-service.yaml [hunk 7]
```diff
@@ -43,14 +44,17 @@ spec:
...
-  {{- range .Values.webserver.service.ports }}
-    - {{ range $key, $val := . }}
-      {{- $key }}: {{ tpl (toString $val) $ }}
-      {{ end }}
+  ports:
+  {{ range .Values.webserver.service.ports }}
+    -
+      {{- range $key, $val := . }}
+      {{ $key }}: {{ tpl (toString $val) $ }}
+      {{- end }}
```
Comment (SayakMukhopadhyay) [inline chart/templates/webserver/webserver-service.yaml:52]: Hey folks...sorry for necroing this PR but I was wondering if anyone has any info about why this change was made. It creates an YAML with an unnecessary line break.
